### DISCUSSION TITLE
Using keccak256 with openpgp

### DISCUSSION THREAD

**Discussion by nolash** [openpgpjs/openpgpjs#1258] [Q&A]
Question (Sorry I don't want to use gitter and didnt find forum or email anywhere).

The hash enum list suggests keccak256 is not implemented.

Is it possible to hook a hash function that the sign method will use?

If not, where would I look to _hack_ something in?

Thanks.

**Comment by twiss** [comment]
Hey :wave: We've started using the Discussions tab as a sort of forum 😊

The draft specification (https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-08#section-9.5) specifies two hash functions in the same family as keccak256, namely SHA3-256 and SHA3-512. I would recommend using those instead, if any.

However, there is no easy way to "hook" in implementations for these in OpenPGP.js v5. So you would have to create a fork to add them. (That might be a good idea anyway, in case we want to merge them into upstream at some point - at the moment I'm not entirely sure whether the added size of these implementations is worth it, compared to SHA2 which is also implemented natively in Web Crypto. Hopefully SHA3 will be too, in the future.)

**Comment by nolash** [comment]
> at the moment I'm not entirely sure whether the added size of these implementations is worth it,

I was thinking more along the lines of a pluggable hash callback. That would not bloat the code, simply enable the caller to provide a hasher of choice if so desired.

**Comment by twiss** [comment]
We have a mechanism to load code on demand, if it's not needed by default, in the lightweight build: e.g. https://github.com/openpgpjs/openpgpjs/blob/fa3c36fc87c8a0b11ff456af4e626a3a56179435/src/crypto/public_key/elliptic/indutnyKey.js#L43

We could do something similar for SHA3, and then for those using the lightweight build, they would only have to load this code if they set the default hash function to SHA3, or if they receive a message that uses it.

I would however prefer not to add the ability to use custom hash functions, as then they won't be interoperable with the wider OpenPGP ecosystem, and this library is after all called OpenPGP.js :)

**Comment by willfarrell** [comment]
[hash-wasm](https://www.npmjs.com/package/hash-wasm) might be a good candidate for a polyfill till the browsers catch up. The SHA3 implementation is ~4kb gzip.
